### PR TITLE
LPD-105501 Render activation and reset forms in the user's language

### PR DESCRIPTION
**Pull request by wanderlast**
https://liferay.atlassian.net/browse/LPD-105501

This supersedes #8094, which was forwarded to brianchandotcom#181776. It is an attempt to fulfill the changes Brian Chan requested there. No reply was posted on that pull request.

## What Is Being Fixed

The account activation email builds its link from the `$password_setup_url$` token, and that link carried a `languageId` parameter that nothing in the portal reads. Request locale is resolved by `PortalImpl.getLocale`, and `languageId` appears nowhere in that chain — it is read in exactly one place, `UpdateLanguageAction` for `/c/portal/update_language`.

The recipient opens the link as a guest, so the user-locale step is skipped and there is no session locale or language cookie on a first visit. Resolution fell through to the company default and the form always rendered in that language. The same URL is built a second time in `sendPassword`, for the reset email behind Forgot Password, carrying no locale parameter at all and having the identical defect.

## How It Is Being Fixed

Both call sites now emit `doAsUserLanguageId`, which `getLocale` already honours on any request and ahead of its fall-through to the company default, so nothing changes on the receiving side. It degrades gracefully: when the user's locale is not among the instance's available locales, resolution continues down the chain to the current behaviour, so no link can break. It also survives the GET-to-POST conversion in `UpdatePasswordAction.resendAsPost`, which copies every request parameter into hidden inputs.

Coverage is integration tests in `UserLocalServiceTest`, which had no coverage of these emails before. They assert the emailed URL carries `doAsUserLanguageId` and no longer carries the inert parameter, and one takes the URL out of the sent email, requests it, and asserts the returned page renders in the user's language.

## Changes Requested By Brian Chan

Two comments on brianchandotcom#181776.

The first was a diff to apply, moving the `screenName` declaration ahead of the `ServiceContext` block in `_addUser`, so the locals read in the order the `addUser` call consumes them. Applied verbatim.

The second flagged the `_postURL` helper as strange. It set `POST` on a link that arrives in an email as a GET, then parsed the returned HTML, sending no body and leaving every parameter in the query string. No integration test in the repository does that: of the six that POST, the five that read the response all send a real body, and the one that does not discards the response. The request also did not match the flow it stands in for, since `resendAsPost` submits the parameters as form fields against the path, with no query string.

`_postURL` now builds an `Http.Options`, adds each parameter with `addPart`, and posts that — the same shape as `EditInGoogleDocsMVCActionCommandTest`, `DLOpenerGoogleDriveManagerTest` and `MCPServerServletTest`, and a match for what the browser actually submits. It also removes the chained `openConnection` that was raised in the earlier review of this branch.

Both `UserLocalServiceTest.testAddUserWithUserLanguageId` and `testSendPasswordWithUserLanguageId` pass against a local bundle after the change.

## PR Check

**pr-check: PASS** — tested on `78588f177a31278ab376265cafa853993bf2bdf8`

| Validation | Result |
| --- | --- |
| Service Builder | PASS |
| Source Format | PASS |
| Service Registration | PASS |
| Transaction Usage | PASS |
| Full Portal Build | NOT VERIFIED |
| Per-Module Compile | NOT VERIFIED |
| Integration Test Compile | PASS |
| Cross-Module Compile | PASS |
| Baseline | PASS (baseline-all + seven Ant projects + 1 changed exporting module) |
| Java Unit Tests | NOT VERIFIED |

Source Format reported no `SourceCheck` violations and applied no fixes, so it made no commit. The yarn side was skipped rather than clean, since the diff contains no js, json, jsp, jspf, scss, ts or tsx file.

Full Portal Build was trimmed at the Pass 1 estimate checkpoint, where the matched set came to roughly 22 minutes, so it established nothing about this branch. Portal core was exercised instead by `ant compile install-portal-snapshots`, and Cross-Module Compile confirmed the installed snapshot jar carried the branch change before compiling against it.

Per-Module Compile verified nothing. The only changed path under `modules` is `src/testIntegration`, which the validation excludes from the deploy set because Integration Test Compile covers it and `-test` modules deploy no runtime bundle, leaving the deploy set empty. Its lockfile check ran and had nothing to check, since the diff touches no `package.json`.

Baseline reported one finding, `EXCESSIVE VERSION INCREASE` on `com.liferay.portal.kernel.test.util` in `portal-test`, lowering `11.0.0` to `10.10.0`. It is inherited rather than this branch's: `git diff liferay-commerce/master..HEAD` on that `packageinfo` is empty, and the branch touches only `portal-impl` and `modules/apps/user/user-test`. The repair was restored and not committed. The branch itself owns no baseline finding, and the Local Version Check passed — the `UserLocalServiceImpl` diff adds and removes no `public` or `protected` line, so no version bump is owed.

Java Unit Tests verified nothing. There is no unit-test counterpart for `UserLocalServiceImpl` anywhere in the tree, and none under `portal-impl/test/unit` or `portal-kernel/test/unit`. The class is covered by the integration test `UserLocalServiceTest` in `apps:user:user-test` instead, which this validation does not run.

### CODE REVIEW

File: modules/apps/user/user-test/src/testIntegration/java/com/liferay/user/service/test/UserLocalServiceTest.java [hunk 1]
```diff
@@ -2008,6 +2100,38 @@ private void _assertUserHasPasswordPolicy(boolean ldapUser, User user)
...
+			String[] values = entry.getValue();
+
+			options.addPart(name, values[0]);
+		}
+
+		options.setLocation(urlString.substring(0, index));
+		options.setPost(true);
+
+		return _http.URLtoString(options);
+	}
```
Comment (smottal) [inline modules/apps/user/user-test/src/testIntegration/java/com/liferay/user/service/test/UserLocalServiceTest.java:2136]: I think this is worst then before...
is not possible to use something like this?
```
private HttpResponse<String> _send(String method) throws Exception {
		return _httpClient.send(
			HttpRequest.newBuilder(
			).uri(
				URI.create(
					"http://localhost:" +
						PortalUtil.getPortalServerPort(false) +
							"/o/.well-known/oauth-protected-resource/mcp")
			).method(
				method, HttpRequest.BodyPublishers.noBody()
			).build(),
			HttpResponse.BodyHandlers.ofString());
	}
```

**Review by smottal — CHANGES_REQUESTED** [review]

**Comment by wanderlast** [comment]
**pr-check: PASS** — tested on `48e0dc4074288bb985a2da177c36140cd759a07f`

| Validation | Result |
| --- | --- |
| Service Builder | PASS |
| Source Format | PASS |
| Service Registration | PASS |
| Transaction Usage | PASS |
| Full Portal Build | PASS |
| Per-Module Compile | NOT VERIFIED |
| Integration Test Compile | PASS |
| Cross-Module Compile | PASS |
| Baseline | PASS (baseline-all + seven Ant projects + 1 changed exporting module) |
| Java Unit Tests | NOT VERIFIED |

Source Format reported no `SourceCheck` violations and applied no fixes, so it made no commit. The yarn side was skipped rather than clean: the diff contains no js, json, jsp, jspf, scss, ts or tsx file, so `skip.node.task` was set and `downloadNode`, `yarnInstall` and the yarn formatter never ran. `-Dvalidate.commit.messages=true` was confirmed to have actually taken effect by rerunning under `ant -v`, which echoed the flag in the `SourceFormatter` argv.

Full Portal Build passed in 5 minutes 48 seconds, recompiling portal-kernel (3722 sources) and portal-impl (2233 sources) from clean with no diagnostics. It covers only one of the two changed files. `modules/apps/user/user-test` carries no `.lfrbuild-portal` marker, so the deploy target's marketplace branch never builds it, and the test file is covered by Integration Test Compile instead.

Per-Module Compile verified nothing. The only changed path under `modules` is `src/testIntegration`, which the validation excludes from the deploy set because Integration Test Compile covers it and `-test` modules deploy no runtime bundle, leaving the deploy set empty. No expansion fired either: the diff removes no `public` or `protected` member, and touches neither `modules/frontend-sdk` nor `modules/node-scripts.config.js`. Its lockfile check ran and had nothing to check, since the diff touches no `package.json`.

Integration Test Compile compiled `apps:user:user-test` with `--rerun`, and the task executed rather than reporting `UP-TO-DATE` or `FROM-CACHE`. `javap` on the resulting class file confirms both new test methods reached it. A `bookmarks-test` control ran first and passed.

Cross-Module Compile resolved one consumer, `apps:user:user-test`, reached through both changed type names, and compiled it with `--rerun`. The deprecated-profile scan covered three marker modules holding 116 Java files under `src/main` and matched neither symbol, so that scan was clean rather than vacuous.

Baseline reported no findings and left the tree clean, so it made no commit. The `EXCESSIVE VERSION INCREASE` on `com.liferay.portal.kernel.test.util` in `portal-test` that the previous run on this branch inherited did not reproduce: this run's `ant all` rebuilt and re-baselined that jar from clean, and the `packageinfo` reads `version 11.0.0` in the tree, in `HEAD`, and on master alike. The Local Version Check passed, since the `UserLocalServiceImpl` diff adds and removes no `public` or `protected` line and so owes no version bump; `com.liferay.portal.service.impl` is not among portal-impl's exported packages in any case.

Java Unit Tests verified nothing. `UserLocalServiceImpl` has no unit-test counterpart anywhere in the tree, while its sibling package under `portal-impl/test/unit/com/liferay/portal/service/impl` holds several others, so the absence is specific to this class rather than a wrong path. The class is covered by the integration test `modules/apps/user/user-test/src/testIntegration/java/com/liferay/user/service/test/UserLocalServiceTest.java`, which this branch extends with the two `doAsUserLanguageId` tests, and which this validation does not run. No suite was run, since there was no class to schedule.

File: modules/apps/user/user-test/src/testIntegration/java/com/liferay/user/service/test/UserLocalServiceTest.java (hunk 1 above)
Comment (wanderlast) [reply modules/apps/user/user-test/src/testIntegration/java/com/liferay/user/service/test/UserLocalServiceTest.java:2136]: I've changed it so it's closer to what you've done. I admittedly wasn't sure in what way to interpret BChan's request since he just mentioned that it was 'strange'. Hopefully this looks better to you?

**Comment by smottal** [comment]
Superseded by https://github.com/liferay-commerce/liferay-portal/pull/8158.
